rpc: replace dprcpool.Conn usage with dprc.Conn

cthumuluru-crdb opened the pull request:
`drpc.Conn` is the connection interface and `drpcpool.Conn` is a wrapper interface to provide connection pool functionality. Currently we use `drpcpool.Conn` across our codebase.

This change replaces all references to `drpcpool.Conn` with the more generic `drpc.Conn`, allowing us to plug in either a plain connection or a pooled connection without touching call sites. In the future, we will be able to introduce connection multiplexing on `drpc.Conn` without touching existing code. Pooled connections can still be used simply by passing in a `drpcpool.Conn` implementation.

Fixes: https://github.com/cockroachdb/cockroach/issues/147199
Epic: CRDB-48923
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147172)
<!-- Reviewable:end -->

shubhamdhama reviewed (commented):
https://github.com/cockroachdb/cockroach/pull/147172/commits/1c1f0eab66038ff5a733e2eb7f4ffbe9e75ad938 LGTM!
Once the stacked commits are removed, I'll approve!

shaikzakiriitm reviewed (commented):
looks good, thanks!

cthumuluru-crdb committed:
rpc: replace dprcpool.Conn usage with dprc.Conn
`drpc.Conn` is the connection interface and `drpcpool.Conn` is a wrapper
interface to provide connection pool functionality. Currently we use
`drpcpool.Conn` across our codebase.

This change replaces all references to `drpcpool.Conn` with the more
generic `drpc.Conn`, allowing us to plug in either a plain connection or
a pooled connection without touching call sites. In the future, we will
be able to introduce connection multiplexing on `drpc.Conn` without
touching existing code. Pooled connections can still be used simply by
passing in a `drpcpool.Conn` implementation.

Fixes: #147199
Epic: CRDB-48923
Release note: none
